src/common/latest/sonarqube/Scanner.ts

```diff
@@ -31,17 +32,34 @@ export default class Scanner {
     return this.isSonarCloud;
   }
 
-  static async downloadScanner(fileUrl: string) {
+  static async downloadScanner(fileUrl: string, name: string, version: string) {
+    let cachedPath;
+
+    const semverCompliantVersion = semver.coerce(version)?.version;
+    if (!semverCompliantVersion) {
+      throw new Error(`Invalid version format: ${version}`);
+    }
+
     try {
-      log(LogLevel.DEBUG, `Downloading scanner from ${fileUrl}`);
+      cachedPath = toolLib.findLocalTool(name, semverCompliantVersion);
+      if (cachedPath) {
+        log(LogLevel.DEBUG, `Found cached scanner at ${cachedPath}`);
+        return cachedPath;
+      }
+      log(LogLevel.DEBUG, `Cache not found for ${name} ${semverCompliantVersion}`);
+      log(LogLevel.INFO, `Downloading scanner from ${fileUrl}`);
       const downloadPath = await toolLib.downloadTool(fileUrl);
       log(LogLevel.DEBUG, `Downloaded: ${fileUrl} file to ${downloadPath}`);
 
       log(LogLevel.DEBUG, `Extracting ${downloadPath}`);
       const unzipPath = await toolLib.extractZip(downloadPath);
       log(LogLevel.DEBUG, `Unzipped file to ${unzipPath}`);
 
-      return unzipPath;
+      // Cache the extracted file
```

Comment:
Tiny NP: Not sure this comment is relevant, code below is self explanatory + there's a debug log which acts as a comment just below